File: backend/ecommerce/src/main/java/com/bookstore/ecommerce/app/domain/Order.java
package com.bookstore.ecommerce.app.domain;

import java.time.Instant;
import java.util.ArrayList;
import java.util.UUID;

import com.bookstore.ecommerce.app.domain.data.Book;
import com.bookstore.ecommerce.app.domain.data.Customer;
import com.bookstore.ecommerce.app.domain.data.OrderItem;
import com.bookstore.ecommerce.app.domain.data.OrderStatus;

import lombok.Getter;

public class Order {
  @Getter
  private final com.bookstore.ecommerce.app.domain.data.Order state;

  public Order(Customer customer, Book[] books) {
    final var id = UUID.randomUUID().toString();

    final var items = new ArrayList<OrderItem>();

    for (final var book : books) {
      items.add(OrderItem.builder()
        .key(new OrderItem.Key(id, book.getId()))
        .bookTitle(book.getTitle())
        .bookSubtitle(book.getSubtitle())
        .bookDescription(book.getDescription())
        .build());
    }

    this.state = com.bookstore.ecommerce.app.domain.data.Order.builder()
      .id(id)
      .number(id)
      .createdAt(Instant.now())
      .status(OrderStatus.Queued.toString())
      .customerName(customer.getName())
      .customerPhone(customer.getPhone())
      .customerDeliveryAddress(customer.getDeliveryAddress())
      .items(items)
      .build();
  }
}

File: order/app/domain/order.go
package domain

import (
	"fmt"
	"time"

	"github.com/thoas/go-funk"
)

type OrderStatus string

const (
	OrderStatusPending   OrderStatus = "Pending"
	OrderStatusAccepted  OrderStatus = "Accepted"
	OrderStatusRejected  OrderStatus = "Rejected"
	OrderStatusCancelled OrderStatus = "Cancelled"
	OrderStatusDelivered OrderStatus = "Delivered"
)

type OrderData struct {
	Id                      string
	CreatedAt               time.Time
	Status                  OrderStatus
	CustomerId              string
	CustomerName            string
	CustomerPhone           string
	CustomerDeliveryAddress string
	Items                   []OrderItem
}

type OrderItem struct {
	OrderId          string
	BookId           string
	BookTitle        string
	BookSubtitle     string
	BookDescription  string
	BookThumbnailUrl string
	Qty              int
}

type Order struct {
	EventSource
	order OrderData
}

func (Order) New(customer Customer, books []Book) *Order {
	customerData := customer.State()

	return &Order{
		order: OrderData{
			Status:                  OrderStatusPending,
			CreatedAt:               time.Now(),
			CustomerId:              customerData.Id,
			CustomerName:            customerData.Name,
			CustomerPhone:           customerData.Phone,
			CustomerDeliveryAddress: customerData.DeliveryAddress,
			Items: funk.Map(books, func(book Book) OrderItem {
				bookData := book.State()

				return OrderItem{
					BookId:           bookData.Id,
					BookTitle:        bookData.Title,
					BookDescription:  bookData.Description,
					BookSubtitle:     bookData.Subtitle,
					BookThumbnailUrl: bookData.ThumbnailUrl,
				}
			}).([]OrderItem),
		},
	}
}

func (order *Order) Cancel() error {
	if order.order.Status != OrderStatusPending &&
		order.order.Status != OrderStatusAccepted {
		return fmt.Errorf(
			"order status is '%s', no cancellation allowed",
			order.order.Status,
		)
	}

	return nil
}
